File: dapur/cek_gambar.py
import os
import re

def scan_unused_images():
    # 👉 PERBAIKAN: Jangan cuma scan 'artikel/', tapi scan root '.'
    # Kita akan memfilter folder mana saja yang mau di-scan di dalam loop
    root_folder = './'
    img_folder = './img/'
    output_file = './img/gambarnganggur.txt'

    if not os.path.exists(img_folder):
        return

    # Ambil semua gambar fisik yang ada di folder img
    all_images = {f for f in os.listdir(img_folder) if f.lower().endswith('.webp')}

    used_images = set()

    # REGEX: Mencari nama file .webp di dalam kode HTML
    webp_pattern = re.compile(r'([^/\\"\']+\.webp)', re.IGNORECASE)

    # Folder yang boleh di-scan (Kategori Mas + folder artikel lama)
    # Kita skip folder sistem agar scan lebih cepat
    skip_folders = {'node_modules', '.git', 'img', 'sementara', 'artikelx', 'mini', 'ext', '.github'}

    for root, dirs, files in os.walk(root_folder):
        # Filter agar tidak masuk ke folder sistem
        dirs[:] = [d for d in dirs if d not in skip_folders]

        for file in files:
            # Hanya scan file HTML
            if file.lower().endswith('.html'):
                file_path = os.path.join(root, file)
                try:
                    with open(file_path, 'r', encoding='utf-8') as f:
                        content = f.read()
                        matches = webp_pattern.findall(content)
                        for match in matches:
                            clean_name = os.path.basename(match)
                            used_images.add(clean_name)
                except Exception as e:
                    print(f"⚠️ Gagal membaca {file_path}: {e}")

    # Logika Matematika Himpunan: (Semua Gambar) - (Gambar yang ditemukan di HTML)
    unused_images = all_images - used_images

    # Pastikan file output bersih dulu sebelum ditulis (overwrite/clean start)
    # Kita pakai 'w' bukan 'a' supaya tidak tumpuk-tumpuk data lama
    if unused_images:
        with open(output_file, 'w', encoding='utf-8') as out:
            for img in sorted(unused_images):
                out.write(img + '\n')
        print(f"✅ V6.9 Scan: Ditemukan {len(unused_images)} gambar nganggur.")
    else:
        # Jika semua terpakai, kosongkan file agar workflow tidak menghapus apapun
        open(output_file, 'w').close()
        print("😎 Aman! Semua file .webp terdeteksi di seluruh folder kategori.")

if __name__ == "__main__":
    scan_unused_images()

File: dapur/cek-gambar.js
import fs from 'node:fs';
import path from 'node:path';
import { execSync } from 'node:child_process';

const IMG_FOLDER = './img/';
const OUTPUT_FILE = './img/gambarnganggur.txt';
const SKIP_FOLDERS = ['node_modules', '.git', 'img', 'sementara', 'artikelx', 'mini', 'ext', '.github'];

/**
 * Fungsi Rekursif untuk ambil SEMUA .webp di dalam img/ sub-sub-sub folder
 */
function getAllPhysicalImages(dir) {
    let results = [];
    if (!fs.existsSync(dir)) return results;

    const list = fs.readdirSync(dir, { withFileTypes: true });
    for (const file of list) {
        const res = path.join(dir, file.name);
        if (file.isDirectory()) {
            results = results.concat(getAllPhysicalImages(res));
        } else if (file.name.toLowerCase().endsWith('.webp')) {
            results.push({
                fullPath: res,
                basename: file.name
            });
        }
    }
    return results;
}

/**
 * Fungsi Rekursif untuk scan referensi di semua file .html
 */
const usedBasenames = new Set();
function walkAndScanHtml(dir) {
    const files = fs.readdirSync(dir, { withFileTypes: true });
    for (const file of files) {
        const fullPath = path.join(dir, file.name);
        if (file.isDirectory()) {
            if (SKIP_FOLDERS.includes(file.name)) continue;
            walkAndScanHtml(fullPath);
        } else if (file.name.endsWith('.html')) {
            try {
                const content = fs.readFileSync(fullPath, 'utf8');
                const matches = content.match(/([^/\\"\']+\.webp)/ig);
                if (matches) {
                    matches.forEach(m => usedBasenames.add(path.basename(m)));
                }
            } catch (e) {
                console.warn(`⚠️ Gagal baca ${fullPath}`);
            }
        }
    }
}

console.log('🚀 Memulai Deep Scan V10.0 (Recursive Img Mode)...');

// 1. Ambil semua gambar fisik (Deep Dive)
const allImages = getAllPhysicalImages(IMG_FOLDER);

// 2. Cari semua yang terpakai di HTML
walkAndScanHtml('./');

// 3. Filter: Cari yang fisiknya ada tapi namanya TIDAK ADA di HTML
const unused = allImages.filter(img => !usedBasenames.has(img.basename));

// 4. Eksekusi
if (unused.length > 0) {
    // Tulis full path ke log supaya kita tahu lokasi persisnya di subfolder mana
    const logContent = unused.map(img => img.fullPath).sort().join('\n') + '\n';
    fs.writeFileSync(OUTPUT_FILE, logContent);
    
    console.log(`✅ Ditemukan ${unused.length} gambar nganggur di kedalaman folder.`);

    let cleanedCount = 0;
    unused.forEach(img => {
        if (fs.existsSync(img.fullPath)) {
            try {
                console.log(`→ Git RM: ${img.fullPath}`);
                // Gunakan quotes "" untuk jaga-jaga jika ada spasi di nama folder/file
                execSync(`git rm -f "${img.fullPath}"`);
                cleanedCount++;
            } catch (err) {
                console.error(`❌ Gagal hapus ${img.fullPath}`);
            }
        }
    });
    console.log(`Successfully cleaned ${cleanedCount} files from various depths.`);
} else {
    fs.writeFileSync(OUTPUT_FILE, '');
    console.log("😎 Aman! Semua gambar di semua subfolder terpakai.");
}
